File: src/main/java/org/example/kalkulationsprogramm/service/ArtikelMatchingService.java
package org.example.kalkulationsprogramm.service;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;

import org.apache.commons.text.similarity.JaroWinklerSimilarity;
import org.example.kalkulationsprogramm.domain.Artikel;
import org.example.kalkulationsprogramm.domain.Lieferanten;
import org.example.kalkulationsprogramm.repository.ArtikelRepository;
import org.example.kalkulationsprogramm.repository.LieferantenRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.AllArgsConstructor;

/**
 * Hilfsservice, der ausgehend von Produktinformationen passende Artikel und Lieferanten vorschlägt.
 */
@Service
@AllArgsConstructor
public class ArtikelMatchingService {

    private final ArtikelRepository artikelRepository;
    private final LieferantenRepository lieferantenRepository;

    private final JaroWinklerSimilarity jaro = new JaroWinklerSimilarity();

    /**
     * Liefert eine Liste von maximal fünf Artikeln, die anhand einer Jaro-Winkler-Ähnlichkeit
     * am besten zu den übergebenen Produktdaten passen.
     */
    @Transactional(readOnly = true)
    public List<Artikel> findeBesteTreffer(String produktname, String produktlinie) {
        List<Artikel> kandidaten = artikelRepository.findAll();
        return kandidaten.stream()
                .sorted(Comparator.comparingDouble(
                        a -> -score(produktname, produktlinie, a.getProduktname(), a.getProduktlinie())))
                .limit(5)
                .toList();
    }

    private double score(String name, String line, String artName, String artLine) {
        return similarity(name, artName) * 0.7 + similarity(line, artLine) * 0.3;
    }

    private double similarity(String a, String b) {
        if (a == null || b == null) return 0;
        return jaro.apply(a.toLowerCase(), b.toLowerCase());
    }

    /**
     * Versucht einen Lieferanten anhand der Domain der übergebenen E-Mail zu finden.
     */
    @Transactional(readOnly = true)
    public Optional<Lieferanten> findeLieferantFuerEmail(String email) {
        if (email == null || !email.contains("@")) return Optional.empty();
        String domain = email.substring(email.indexOf('@') + 1).trim().toLowerCase();
        List<Lieferanten> matches = lieferantenRepository.findByEmailDomain(domain);
        return matches.isEmpty() ? Optional.empty() : Optional.of(matches.getFirst());
    }

    /**
     * Speichert eine E-Mail-Adresse für den angegebenen Lieferanten, falls sie noch nicht vorhanden ist.
     */
    @Transactional
    public void merkeLieferantenEmail(Lieferanten lieferant, String email) {
        if (lieferant == null || email == null) return;
        String cleaned = email.trim();
        if (lieferant.getKundenEmails().stream().map(e -> e == null ? null : e.trim())
                .noneMatch(e -> cleaned.equalsIgnoreCase(e))) {
            lieferant.getKundenEmails().add(cleaned);
            lieferantenRepository.save(lieferant);
        }
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/ArtikelMatchingService.kt
package org.example.kalkulationsprogramm.service

import org.apache.commons.text.similarity.JaroWinklerSimilarity
import org.example.kalkulationsprogramm.domain.Artikel
import org.example.kalkulationsprogramm.domain.Lieferanten
import org.example.kalkulationsprogramm.repository.ArtikelRepository
import org.example.kalkulationsprogramm.repository.LieferantenRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.Optional

@Service
class ArtikelMatchingService(
    private val artikelRepository: ArtikelRepository,
    private val lieferantenRepository: LieferantenRepository,
) {
    private val jaro = JaroWinklerSimilarity()

    @Transactional(readOnly = true)
    fun findeBesteTreffer(produktname: String?, produktlinie: String?): List<Artikel> =
        artikelRepository.findAll()
            .sortedByDescending { artikel ->
                score(produktname, produktlinie, readString(artikel, "getProduktname"), readString(artikel, "getProduktlinie"))
            }
            .take(5)

    @Transactional(readOnly = true)
    fun findeLieferantFuerEmail(email: String?): Optional<Lieferanten> {
        if (email == null || !email.contains("@")) {
            return Optional.empty()
        }
        val domain = email.substring(email.indexOf('@') + 1).trim().lowercase()
        return lieferantenRepository.findAll().stream()
            .filter { lieferant ->
                lieferant.kundenEmails
                    .map { it.trim().lowercase() }
                    .any { it.endsWith(domain) }
            }
            .findFirst()
    }

    @Transactional
    fun merkeLieferantenEmail(lieferant: Lieferanten?, email: String?) {
        if (lieferant == null || email == null) {
            return
        }
        val cleaned = email.trim()
        if (lieferant.kundenEmails.map { it.trim() }.none { cleaned.equals(it, ignoreCase = true) }) {
            lieferant.kundenEmails.add(cleaned)
            lieferantenRepository.save(lieferant)
        }
    }

    private fun score(name: String?, line: String?, artName: String?, artLine: String?): Double =
        similarity(name, artName) * 0.7 + similarity(line, artLine) * 0.3

    private fun similarity(a: String?, b: String?): Double {
        if (a == null || b == null) {
            return 0.0
        }
        return jaro.apply(a.lowercase(), b.lowercase())
    }

    private fun readString(target: Any, getterName: String): String? =
        runCatching { target.javaClass.getMethod(getterName).invoke(target) as? String }.getOrNull()
}
